### PR TITLE
Ubuntu test

### LINKED ISSUE #4: Build Spigot + basic setup
- ~~[x] Build Spigot~~
  ~~curl buildtools -> tempfs; build -> move jar to /spigot (buildtools are auto-truncated from tempfs once no longer used, kinda?)~~
- [x] get spigot
- [x] get coreProtect
- [x] get DynMap
  
- [x] Handle Start
  - [x] accept eula
  Currently the EULA eula.txt is set using script
  But we found out we can simply pass it as a parameter to spigot.jar when launching
  Note that the ELUA is a legally binding document.
- [x] Handle Restart
- [x] Handle Stop

- [x] SSH
  OpenSSH Server
- [x] User and Permissions
  pingu
  no root

### LINKED ISSUE #8: Minecraft Server Image
Server Type: Spigot
Server Version: ?? (using Latest)

- OS: ??  (using jdk-Alpine)
  Configure Image 
  - [ ] User and Permissions #4 
  - [ ] ssh #4 
  - [ ] iptables
  - [ ] hardening?

- Core Server
  - [ ] Build spiggot.jar #4 
  - Configure spiggot #4 
    - [ ] spiggot preferences  #4
    - [ ] vanilla preferences  #4
    - [ ] bukkit preferences  #4
    - [ ] server icon (optional)
  - [ ] Handles Startup #4
  - [ ] Handles Restart #4 
  - [ ] Handles Stop #4 

- Plugins
  - [ ] DynMap
  - [ ] CoreProtect
  
- Documentation
  - [ ] For all how all checkboxes above are done
  - [ ] But also for relevant files/folders and paths (where are things and how are they used)

### LINKED ISSUE #9: minecraft config files
configurations for:

- [x] server.properties
- [x] bukkit config
- [x] spigot-config
- [x] server-icon

### PR DESCRIPTION
**Pull request by ZKemstedt**
Spigot: Builds Runs Stops ✅ 

DynMap: Works!
CoreProtect Added - Not tested

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,16 @@
...
+
+# old local dev files
+mc-conf
+mc-world
+mc
+old
+local
 
-# nobody is stupid enough to throw their private ones in here, riiiiight?
-super_secret_key_dont_look
```
Comment (Nisserino) [inline .gitignore:5]: I can finally share my super_secret_key WOOH

File: Dockerfile [hunk 2]
```diff
@@ -1,42 +1,68 @@
 # syntax=docker/dockerfile:1
```
Comment (Nisserino) [inline Dockerfile:1]: Remove comments since we no longer go for multi-stage(/layer) build

**Comment by ZKemstedt** [comment]
closes #4

File: Dockerfile [hunk 3]
```diff
@@ -1,42 +1,68 @@
...
+
+# Add a penguin for running the server
+RUN groupadd -g 1000 pingu \
+    && useradd --no-log-init -m -u 1000 -g pingu pingu \
+    && echo "pingu:skt" | chpasswd \
+    && chown pingu:pingu /home/pingu
 
 # SSH key setup
+RUN mkdir /run/sshd
+RUN mkdir /home/pingu/.ssh
```
Comment (Nisserino) [inline Dockerfile]: Can make it a one-liner `RUN mkdir /run/sshd /home/pingu/.ssh`

File: Dockerfile [hunk 4]
```diff
@@ -1,42 +1,68 @@
...
+# Move conf over
+RUN mkdir /tmp/conf
+COPY conf /tmp/conf
+COPY scripts/* /
 
-# Temporarily disabled because I'm really frustrated about having to
-# remove entries from my hosts file every time I rebuild the image
-#RUN rm -rf /etc/ssh/ssh_host_rsa_key /etc/ssh/ssh_host_dsa_key
+# dos2unix fixes line endings for scripts
+RUN dos2unix /start* && chmod +x /start*
```
Comment (Nisserino) [inline Dockerfile]: Missing out on utils.sh
proposed change: `RUN dos2unix /*.sh && chmod +x /*.sh`

File: scripts/start.sh [hunk 5]
```diff
@@ -1,24 +1,3 @@
...
-  echo -e "${YELLOW}Initialize server files...${NC}"
-  /scripts/runserver.sh
-fi
-
-echo -e "${YELLOW}Setting EULA to true...${NC}"
-sed -i 's/false/true/g' eula.txt
-
-echo -e "${YELLOW}Starting server...${NC}"
-/scripts/runserver.sh
+exec java -Xms2G -Xmx8G -jar /data/spigot.jar
```
Comment (Nisserino) [inline scripts/start.sh]: Use environment variables for `-Xms` & `-Xmx`

File: scripts/utils.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+function writeEula {
+  if ! echo "# Generated via Docker
+# $(date)
+eula=${EULA,,}
```
Comment (Nisserino) [inline scripts/utils.sh:6]: What does `,,` do in `eula=${EULA,,}`

File: scripts/utils.sh [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+" >/data/eula.txt; then
+    log "ERROR: unable to write eula to /data. Please make sure attached directory is writable by uid=${UID}"
+    exit 2
+  fi
+}
+
+
+function downloadSpigot {
+  if [[ ${VERSION^^} = LATEST ]]; then
```
Comment (Nisserino) [inline scripts/utils.sh:15]: What does `^^` do in `${VERSION^^}`

**Review by Nisserino — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 8]
```diff
@@ -1,42 +1,68 @@
...
+
+# #NOTE otherwise...
+# RUN mkdir /tmp/build /spg && cd /spg \
+#     && curl -o /tmp/build/BuildTools.jar \
+#     https://hub.spigotmc.org/jenkins/job/BuildTools/lastSuccessfulBuild/artifact/target/BuildTools.jar \
+#     && java -jar /tmp/build/BuildTools.jar --output-dir .
 
-# Create an igloo where it stores server files
-RUN mkdir -m 770 /data && \
-    chown pingu:pingu /data
```
Comment (Nisserino) [inline Dockerfile]: Remove comments since we no longer go for multi-stage(/layer) build

**Review by Nisserino — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (ZKemstedt) [reply .gitignore:5]: ❤️

**Review by Nisserino — DISMISSED** [review]
Wooh

**Review by Nisserino — DISMISSED** [review]
Prod nästa!

**Comment by ZKemstedt** [comment]
Spigot works
Dynmap works

needs testing:
Dynmap DB (mysql)
CoreProtect

File: conf/server.properties [hunk 9]
```diff
@@ -0,0 +1,50 @@
+#Minecraft server properties
+#Sat Sep 11 17:53:29 GMT 2021
+enable-jmx-monitoring=false
+rcon.port=25575
+gamemode=survival
+enable-command-block=false
+enable-query=false
```
Comment (ZKemstedt) [inline conf/server.properties:7]: What is `query`?

File: conf/server.properties [hunk 10]
```diff
@@ -0,0 +1,50 @@
+#Minecraft server properties
+#Sat Sep 11 17:53:29 GMT 2021
+enable-jmx-monitoring=false
```
Comment (ZKemstedt) [inline conf/server.properties:3]: what is `enable-jmx-monitoring`?

File: conf/server.properties [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+broadcast-console-to-ops=true
+spawn-npcs=true
+spawn-animals=true
+snooper-enabled=true
+function-permission-level=2
+text-filtering-config=
+spawn-monsters=true
+enforce-whitelist=false
+resource-pack-sha1=
+spawn-protection=0
```
Comment (ZKemstedt) [inline conf/server.properties]: should probably set to 1

File: conf/spigot.yml [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+config-version: 12
+settings:
+  debug: false
+  bungeecord: false
+  sample-count: 12
+  player-shuffle: 0
+  user-cache-size: 1000
+  save-user-cache-on-stop-only: false
+  moved-wrongly-threshold: 0.0625
+  moved-too-quickly-multiplier: 10.0
```
Comment (ZKemstedt) [inline conf/spigot.yml:22]: Hopefully these shouldn't do anything as the `"allow flying"` option in `server.properties` is set to `true`

File: scripts/start.sh [hunk 13]
```diff
@@ -1,24 +1,3 @@
...
-  echo -e "${YELLOW}Initialize server files...${NC}"
-  /scripts/runserver.sh
-fi
-
-echo -e "${YELLOW}Setting EULA to true...${NC}"
-sed -i 's/false/true/g' eula.txt
-
-echo -e "${YELLOW}Starting server...${NC}"
-/scripts/runserver.sh
+exec java -Xms${XMS} -Xmx${XMX} -XX:+UseG1GC -XX:+ParallelRefProcEnabled -XX:MaxGCPauseMillis=200 -XX:+UnlockExperimentalVMOptions -XX:+DisableExplicitGC -XX:+AlwaysPreTouch -XX:G1NewSizePercent=30 -XX:G1MaxNewSizePercent=40 -XX:G1HeapRegionSize=8M -XX:G1ReservePercent=20 -XX:G1HeapWastePercent=5 -XX:G1MixedGCCountTarget=4 -XX:InitiatingHeapOccupancyPercent=15 -XX:G1MixedGCLiveThresholdPercent=90 -XX:G1RSetUpdatingPauseTimePercent=5 -XX:SurvivorRatio=32 -XX:+PerfDisableSharedMem -XX:MaxTenuringThreshold=1 -Dusing.aikars.flags=https://mcflags.emc.gs -Daikars.new.flags=true -jar /data/spigot.jar nogui
```
Comment (ZKemstedt) [inline scripts/start.sh]: split args to multiple lines for readability

File: scripts/utils.sh (hunk 6 above)
Comment (ZKemstedt) [reply scripts/utils.sh:6]: converts the value of  $EULA to lowercase

File: scripts/utils.sh (hunk 7 above)
Comment (ZKemstedt) [reply scripts/utils.sh:15]: converts the value of $VERSION to uppercase

File: conf/plugins/dynmap/configuration.txt [hunk 14]
```diff
@@ -0,0 +1,501 @@
...
+# Disables Webserver portion of Dynmap (Advanced users only)
+disable-webserver: false
+
+# Enable/disable having the web server allow symbolic links (true=compatible with existing code, false=more secure (default))
+allow-symlinks: true
+
+# Enable login support
+login-enabled: false
+# Require login to access website (requires login-enabled: true)
+login-required: false
```
Comment (ZKemstedt) [inline conf/plugins/dynmap/configuration.txt:346]: How does login to dynmap actually work?

File: conf/plugins/dynmap/configuration.txt [hunk 15]
```diff
@@ -0,0 +1,501 @@
...
+# Default selection when map page is loaded
+defaultzoom: 0
+defaultworld: world
+defaultmap: flat
+# (optional) Zoom level and map to switch to when following a player, if possible
+#followzoom: 3
+#followmap: surface
+
+# If true, make persistent record of IP addresses used by player logins, to support web IP to player matching
+persist-ids-by-ip: true
```
Comment (ZKemstedt) [inline conf/plugins/dynmap/configuration.txt:424]: disable for dev
enable for prod

File: conf/server.properties (hunk 11 above)
Comment (Nisserino) [reply conf/server.properties]: Yeah, I'll update it
Apparently it protects a 2x+n square, and not n chunks as I thought

File: conf/spigot.yml (hunk 12 above)
Comment (Nisserino) [reply conf/spigot.yml:22]: Fingers crossed

File: conf/server.properties (hunk 9 above)
Comment (Nisserino) [reply conf/server.properties:7]: from minecraft.fandom.com
> Enables GameSpy4 protocol server listener. Used to get information about server. 
I think we should just keep it as false

File: conf/server.properties (hunk 10 above)
Comment (Nisserino) [reply conf/server.properties:3]: It has something to do with java and beans.

> Exposes an MBean with the Object name net.minecraft.server:type=Server and two attributes averageTickTime and tickTimes exposing the tick times in milliseconds.

> In order for enabling JMX on the Java runtime you also need to add a couple of JVM flags to the startup as documented [here](https://docs.oracle.com/javase/8/docs/technotes/guides/management/agent.html)

Not really sure what it does, but I don't think it's something worth digging into that much

File: scripts/setup.sh [hunk 16]
```diff
@@ -1,15 +1,17 @@
 #!/bin/bash
 
+# On first start after build
 if [ ! "$(ls -A /data)" ]; then
+    # Move over configuration files
+    mv /tmp/conf/* /data/
+    rm -rf /tmp/conf
```
Comment (ZKemstedt) [inline scripts/setup.sh]: Actually nvm this wont work due to our persistent volumes.
Instead we should check for the existance of the /tmp/conf directory and act upon that.

the problem with mv is that it cannot rewrite target directories if they are not empty so for this reason we will have to move over the files individually

Comment (ZKemstedt) [reply scripts/setup.sh]: Fixed

**Review by Nisserino — APPROVED** [review]
Woopdiewoop, det börjar ju faktiskt likna något ballt nu

**Review by CoolAutumnRain — APPROVED** [review]

**Pull request merged**